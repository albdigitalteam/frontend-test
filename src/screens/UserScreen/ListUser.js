import React, { memo, useCallback } from 'react';
import { FlatList, RefreshControl } from 'react-native';
import styled from 'styled-components/native';

import PropTypes from 'prop-types';

import colors from '@/theme/colors';

import CardUser from '@/components/CardUser';

const ListUser = ({ users, isLoading, onRefresh }) => {
  const renderItem = ({ item: { id, name } }) => {
    return <Card id={id} name={name} />;
  };

  const reflashControl = useCallback(
    () => (
      <RefreshControl
        progressViewOffset={32}
        tintColor={colors.PRIMARY}
        colors={[colors.PRIMARY]}
        refreshing={isLoading}
        onRefresh={onRefresh}
      />
    ),
    [isLoading, onRefresh]
  );

  return (
    <StyledContainer>
      <FlatList
        data={users}
        keyExtractor={item => item.id}
        showsVerticalScrollIndicator={false}
        refreshControl={reflashControl()}
        renderItem={renderItem}
        refreshing={isLoading}
      />
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  flex: 1;
  background-color: ${colors.COLOR_GRAY};
`;

const Card = styled(CardUser)`
  margin-top: 10px;
`;
ListUser.defaultProps = {
  users: [],
  isLoading: false,
  onRefresh: () => {}
};

ListUser.propTypes = {
  isLoading: PropTypes.bool,
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
      username: PropTypes.string,
      email: PropTypes.string
    })
  ),
  onRefresh: PropTypes.func
};

export default memo(ListUser);
