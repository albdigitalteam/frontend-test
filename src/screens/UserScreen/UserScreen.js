import React, { useEffect, useCallback } from 'react';

import styled from 'styled-components/native';

import { useDispatch, useSelector } from 'react-redux';

import Label from '@/components/Label';
import colors from '@/theme/colors';

import { LOADING_USERS } from '@/store/slices/userSlice';

import ListUser from './ListUser';

const UserScreen = () => {
  const dispatch = useDispatch();
  const { usersData, isLoading, errorMessage } = useSelector(
    ({ users }) => users
  );

  const loadingData = useCallback(() => {
    dispatch(LOADING_USERS());
  }, [dispatch]);

  useEffect(() => {
    loadingData();
  }, [loadingData]);

  return (
    <StyledContainer>
      {errorMessage ? (
        <Label
          textAlign='center'
          fontWeight={400}
          fontSize={12}
          marginBottom={8}
          color={colors.DANGER}
        >
          {errorMessage}
        </Label>
      ) : (
        <ListUser
          users={usersData}
          isLoading={isLoading}
          onRefresh={loadingData}
        />
      )}
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  flex: 1;
  padding-left: 8px;
  padding-right: 8px;
  margin-bottom: 10px;
`;

export default UserScreen;
