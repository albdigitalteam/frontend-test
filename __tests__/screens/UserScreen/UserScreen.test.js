import 'react-native';
import React from 'react';
import {
  matchSnapshotWithTheme,
  renderWithTheme
} from '@/helpers/functions/testing';

import UserScreen from '@/screens/UserScreen';

describe('tests in UserScreen Screen', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(<UserScreen />);
  });

  test('renders correctly', () => {
    renderWithTheme(<UserScreen />);
  });
});
