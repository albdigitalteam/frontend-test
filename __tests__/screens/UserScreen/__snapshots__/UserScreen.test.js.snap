// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in UserScreen Screen matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 10,
        "paddingLeft": 8,
        "paddingRight": 8,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#F3F3F3",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          colors={
            Array [
              "#25265E",
            ]
          }
          onRefresh={[Function]}
          progressViewOffset={32}
          refreshing={false}
          tintColor="#25265E"
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
  </View>
</View>
`;
